File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    visited = [[False] * len(maps[0]) for i in range(len(maps))]
    print(visited)
    
    q = deque()
    dx = [-1,1,0,0]
    dy= [0,0,-1,1]
    q.append([0,0,0])
    
    while q :
        movelist = q.popleft()
        x = movelist[0]
        y = movelist[1]
        cnt = movelist[2]
        
        if (x == len(maps)-1) and (y == len(maps[0])-1) :
            return cnt+1
        
        for i in range(4) :
            nx = x+dx[i]
            ny = y+dy[i]
            
            if 0<=nx<len(maps) and 0<=ny<len(maps[0]) and visited[nx][ny] == False:
                #방문안했을때
                if maps[nx][ny] == 1 :
                    visited[nx][ny] = True
                    q.append([nx,ny,cnt+1])
                    
    return -1
                    
      

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    public int [] dx = {0,1,0,-1};
    public int [] dy = {1,0,-1,0};
    
    public int solution(int[][] maps) {
        int answer = 0;
        Queue<int [] > q = new LinkedList<>();
        
        boolean [][] visited = new boolean [maps.length][maps[0].length];
        
        int [] firstarr = {0,0,1}; //x,y,cnt
        q.add(firstarr);
        visited[0][0] = true;
        while (!q.isEmpty())
        {
            int [] now = q.poll();
            int x = now[0];
            int y = now[1];
            int cnt = now[2];
            
            if (x == maps.length-1 && y == maps[0].length-1)
            {
                return cnt;
            }
            
            for (int i = 0 ; i<4; i ++)
            {
                int nx = x+dx[i];
                int ny = y+dy[i];
                
                if (0<=nx && nx<maps.length && 0<=ny && ny<maps[0].length)
                {
                    if (!visited[nx][ny] && maps[nx][ny]==1)
                    {
                        visited[nx][ny] = true;
                        int [] newarr = {nx,ny,cnt+1};
                        q.add(newarr);
                    }
                }
            }
        }
        
        return -1;
    }
}